### DISCUSSION TITLE
PyMacNet within Maccor architecture

### DISCUSSION THREAD

**Discussion by samzitara** [battery-data-alliance/pymacnet#8] [Q&A]
I’d like to learn more about the system architecture of the Maccor cyclers and how PyMacNet fits into the puzzle. It’s difficult to do that without having access to a Maccor cycler. Is there any other documentation available to help me understand this?
Based on PyMacNet and the Maccor website, my current understanding is shown below:
![image](https://user-images.githubusercontent.com/101226209/224522430-504904ca-f363-4ce6-a9b2-a61be1135e96.png)

**Comment by chintanp** [comment]
@samzitara what exactly is it that you are trying to achieve? Depending on the requirements and cycler configuration and available interfaces (CAN, SMBus), different solutions can be attempted. We have access to a shared user facility where you can rent cyclers for your experiments and we can help with setting that up.

**Reply by SamWilson22** [reply]
@chintanp Thank you for following up. I'm trying to learn about the system architecture of Maccor and see how it compares to cyclers I'm more familiar with (Arbin and Neware). I'd like to see where there are opportunities to contribute to PyMacNet, either by adding functionally or by generalizing it to work with other cycler types.  

Renting a Maccor cycler from a shared user facility sounds promising. I'd like to see if there's some other initial options available before spending money. 

A block diagram of the default setup for Maccor, without PyMacNet, CAN, or SMBus, would be a good starting point.

**Reply by chintanp** [reply]
@tanevitt any comments ?

**Reply by tanevitt** [reply]
Thank you for your interest in PyMacNet, @SamWilson22! A subtle distinction that may not be clear from the existing PyMacNet documentation is that PyMacNet actually interfaces with the Maccor test software and not directly with the cycler. To make this clear in your diagram you could add a box for “Maccor Test Software” between “MacNet” and “Maccor Cycler”. The Procedure file also lives within the Maccor test software. You could also add a “MacNet” box between “PyMacNet” and “MaccorSpoofer” as this is how the two communicate.

Regarding your quest to learn more about Maccor system architecture, Maccor and Arbin cyclers largely follow the same design paradigm. If you’re already familiar with Arbin then learning the Maccor system will not be too difficult. 

As far as expanding PyMacNet to include additional features, any of the following would be great:

- Add methods for reading Version Info (1,1), System Info (1,2), Channel Specification (3,1), and/or Calibration Dates (3,2). The specifics for these messages can be found in the `docs/[macnet_from_maccor_help.pdf](https://github.com/BattGenie/pymacnet/blob/main/docs/macnet_from_maccor_help.pdf)` file. These would be relatively easy to add and good first projects. 
- Add method to Check test start on channel (6,11) to see if a test can start on a channel and report back the message to the user. This would also be relatively easy. 
- Add methods for Suspending (6,3) and Resuming (6,4) tests. These would be slightly harder as they would require testing and handling of edge cases (what happens when trying to resume a non-existing test?) 
- Add method to support Advanced Start (6,13). This would allow for restarting completed tests or starting new tests at specific steps or cycles. This would definitely be the most difficult as there are many potential corner cases that would need to be addressed

Expanding PyMacNet would be made more difficult by not having access to a Maccor cycler, but certainly not impossible. I’d also be happy to test new functionality on our Maccor cycler as it is added. 

I love the idea of expanding PyMacNet to be cycler agnostic (what would we rename it to?!), but this would be a large project. Arbin has their Console TCP Interface (CTI) which is analogous to MacNet, but I have not used it (yet) and I’m not sure how close the feature set is to MacNet. Do you have any experience with CTI? 

Again, thanks for your interest and please let me know if you have any other questions or feedback.
